### DISCUSSION TITLE
A quick guide to getting started with OC

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15264] [Ideas]
I'm continuing my learning path with Orchard Core and have some feedback that could be useful for the project. I've watched the [dojo course 3](https://www.youtube.com/watch?v=aSNRH-BDAdY&list=PLuskKJW0FhJcSX7j0Bd-1X5hq3dgCtYwO) on YouTube and explored [the code](https://github.com/Lombiq/Orchard-Training-Demo-Module?tab=readme-ov-file#how-to-start), but now, **I feel quite lost** 🫠

I've obtained a lot of information on **creating a module** and **handling events**, including **writing to the database**, I installed a clean Orchard on a Web App and I'm unsure (lost) about where to start. For instance:

1. Should I use a Blog/Agency recipe for creating a custom website, or would it be better to start with the blank recipe? 
2. What's the best practice to create my own theme and make it reusable? 
3. My website will feature two types of carousels, should I treat these as Orchard modules to be reused in other projects, or is it better to incorporate them into my theme? 🤔

Upon some search on Google, I discovered that **others have shared my confusion**. For example, this person has written an [excellent guide](https://egoplacebo.medium.com/theming-orchard-core-d78c5cab1336) on theming Orchard Core from scratch, based on notes from a [five year old video](https://www.youtube.com/watch?v=wtAIgn4gYXc), which might still be relevant today (maybe needs to be updated). Unfortunately the guide is not finished.

I believe a beginner's guide to Orchard Core, **_with short tips on best practices for "Content Types"/"Content Parts" and theme creation, is crucial_**. Once familiar with the basics, developers can explore Orchard's deeper functionalities.

What are your thoughts?

Thank you

**Comment by EgoPlacebo** [comment]
As the person mentioned as having started writing a guide, thanks for the feedback :) That should help inspire me to continue! 
I do intend to continue when I have time to figure out exactly how to move forward and note my progress. (I've currently gotten a little stuck in my attempts to create a 'person' module and control the display of text & media fields in custom parts etc.)

I agree it would be wonderful to have more step-by-step tutorials for _reusable_ theme development for those trying to learn - this lack is what inspired me to attempt to write a guide following my own progress. I get the impression that we are expected to learn primarily by reading and exploring the code but, for myself at least, this is not an ideal method for getting started. 

In any event, more resources are likely to increase adoption, which is desirable, surely?

FWIW, here are my OC Novice thoughts on your questions :-P 
1. I think whether you use Blog/Agency or roll your own is a difficult question to answer as it would depend on how complex your needs are - whether these provide what you need.
 2. Obviously I am still working on this myself lol!
 3. Personally, I would develop the carousels as modules, because being able to reuse can only be a positive! In theory you could develop one module which could provide several types of carousels using settings within the module? 

Also, if you do happen to find any good resources not already mentioned, please do pass them along!

**Reply by MarGraz** [reply]
Thank you for your suggestion. 

1. In the end, I decided to start with a "blank recipe" to see how difficult it is to start from scratch and get something simple working;
2. 😁 me too;
3. Probably, I will first add the carousel to the page. Then, once I have a working version, I will create a feature branch to migrate the feature into a module.

Yes, if I find new guides, I will definitely share them here 🙂

**Comment by Piedone** [comment]
My take:

1. I recommend having your own setup recipe (usually in your site's theme). This might be based on, or even executing one or more of the built-in recipes too, but then you can also add your own content/configuration. Using [Auto Setup](https://docs.orchardcore.net/en/latest/docs/reference/modules/AutoSetup/) you can also run the setup with your recipe right at app start.
2. You can take a look at https://github.com/Lombiq/Orchard-Base-Theme/ for inspiration. But generally, if by "reusable" you mean usable on many sites as their site theme, then apart from implementing a generic enough look and feel, I'd add some configuration points. E.g., for configuring a header image or colors (configuration can be done the same way as with modules, like with site settings or [configuration providers](https://docs.orchardcore.net/en/latest/docs/reference/core/Configuration/). However, at this point I wouldn't worry about this much, and rather built the theme. Then, once you actually want to reuse it, you'll have specific problems that you can tackle.
3. Technically, you can do either and they'll work the same. Separate modules are good if you indeed want reusability (especially in different apps/solutions, but at least among multiple tenants), or if your theme would otherwise become huge and you want to structure your code.

**Reply by MarGraz** [reply]
Thank you for your reply.

1. Yes, having a setup recipe is a good approach in terms of reusability and quick startup. However, my question was more about which recipe is the best for creating my own theme and getting it to work in a few days, especially now that I'm a newbie. I've decided to start with a "blank recipe" to see how difficult it is to start from scratch and get something simple working;
2. Yes, the idea is to have a "generic" template that I can install and use on several websites, with features like carousels or contact forms. Of course, then I can work on specific problems. Okay, I will keep an eye on the [Lombiq Orchard Base Theme](https://github.com/Lombiq/Orchard-Base-Theme) and follow it as a best practice example;
3. Okay, clear. I think I will first add the carousel to the page. Then, once I have a working version, I will create a feature branch to migrate the feature into a module.

**I believe a step by step quick guide is necessary to provide users with some best practices for starting a website from scratch.** Of course, once familiar with the basics, developers can explore Orchard's deeper functionalities.

@Piedone, in Lombiq are you planning to work on it in the near future? I'm also writing some notes while [watching this video](https://www.youtube.com/watch?v=wtAIgn4gYXc), as @EgoPlacebo has done, but it's becoming a job 🤣. Of course, it's useful for me to share them with others.

**Reply by Piedone** [reply]
We don't have any plans for further tutorials at the moment. The Dojo Course 3 provides you with a step by step guide in video form, the Training Demo does this for writing code, and our new [Walkthroughs module](https://github.com/Lombiq/Orchard-Walkthroughs) provides interactive walkthroughs for the admin UI.

While I appreciate there's no tutorial that'd cover this specific need of yours, and there is always room for improvement, I think after completing the above (and consulting the docs) you should possess the necessary information to be able to assess these questions.

**Comment by rwxzig** [comment]
It has a steep learning curve, but once you get the gist of it, it starts to become more familiar. However I still have yet to learn module development, and incorporate regular .net mvc code to Orchard module.
